### PR TITLE
k8s: Bump k8s libraries to v0.30.0

### PR DESCRIPTION
**Pull request by RamLavi**
### What this PR does
This PR bumps k8s vendor to v0.30.0 on go.mod files.
It also performs needed changes in order to comply with the new vendor libraries.

Note: This PR follows the [update-k8s-dependencies.](https://github.com/kubevirt/kubevirt/blob/main/docs/update-k8s-dependencies.md) manual.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
Bump k8s deps to 0.30.0
```

### CODE REVIEW

File: pkg/virt-api/api.go [hunk 1]
```diff
@@ -780,7 +780,7 @@ func (app *virtAPIApp) Compose() {
 }
 
 func (app *virtAPIApp) ConfigureOpenAPIService() {
-	config := openapi.CreateConfig()
+	config := openapi.CreateV3Config()
```
Comment (xpivarc) [inline pkg/virt-api/api.go:783]: Why is this needed?

File: pkg/virt-controller/leaderelectionconfig/config.go [hunk 2]
```diff
@@ -39,7 +39,7 @@ func DefaultLeaderElectionConfiguration() Configuration {
 		LeaseDuration: metav1.Duration{Duration: DefaultLeaseDuration},
 		RenewDeadline: metav1.Duration{Duration: DefaultRenewDeadline},
 		RetryPeriod:   metav1.Duration{Duration: DefaultRetryPeriod},
-		ResourceLock:  resourcelock.EndpointsLeasesResourceLock,
+		ResourceLock:  resourcelock.LeasesResourceLock,
```
Comment (xpivarc) [inline pkg/virt-controller/leaderelectionconfig/config.go:42]: Why?

File: staging/src/kubevirt.io/client-go/go.mod [hunk 3]
```diff
@@ -1,6 +1,8 @@
 module kubevirt.io/client-go
 
-go 1.21
+go 1.22.0
```
Comment (xpivarc) [inline staging/src/kubevirt.io/client-go/go.mod:3]: Why?

File: api/openapi-spec/swagger.json [hunk 4]
```diff
@@ -129,60 +129,28 @@
...
-       "in": "query"
+       "$ref": "#/parameters/continue-tuthsW5V"
       },
       {
-       "uniqueItems": true,
-       "type": "string",
-       "description": "A selector to restrict the list of returned objects by their fields. Defaults to everything.",
-       "name": "fieldSelector",
-       "in": "query"
+       "$ref": "#/parameters/fieldSelector-xIcQKXFG"
```
Comment (oshoval) [inline api/openapi-spec/swagger.json:135]: swagger seems wrong

File: staging/src/kubevirt.io/client-go/go.mod (hunk 3 above)
Comment (RamLavi) [reply staging/src/kubevirt.io/client-go/go.mod:3]: I think I saw a runtime log saying that it needed it.. will try to remove it back and see if I get it again.

File: pkg/virt-controller/leaderelectionconfig/config.go (hunk 2 above)
Comment (RamLavi) [reply pkg/virt-controller/leaderelectionconfig/config.go:42]: I that const is not public anymore. The const I used `LeasesResourceLock` is not the same though , so I want to see if this option fails..

Comment (RamLavi) [reply pkg/virt-controller/leaderelectionconfig/config.go:42]: See [here](https://github.com/kubernetes/client-go/blob/35cab326ad02285637a370fcb5a555eb85d1c71a/tools/leaderelection/resourcelock/interface.go#L183)

Comment (RamLavi) [reply pkg/virt-controller/leaderelectionconfig/config.go:42]: But for some reason the `virt-controller` endpoint id [gone](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11840/pull-kubevirt-e2e-k8s-1.30-sig-compute/1786017434541494272#1:build-log.txt%3A11560)... trying to figure out why

File: staging/src/kubevirt.io/client-go/go.mod (hunk 3 above)
Comment (RamLavi) [reply staging/src/kubevirt.io/client-go/go.mod:3]:
```
go: module ./staging/src/kubevirt.io/client-go requires go >= 1.22.0; switching to go1.22.2
```

File: pkg/virt-api/api.go (hunk 1 above)
Comment (RamLavi) [reply pkg/virt-api/api.go:783]: At the time it looks like it was necessary but on second look I think it can be removed. DONE

File: api/openapi-spec/swagger.json (hunk 4 above)
Comment (RamLavi) [reply api/openapi-spec/swagger.json:135]: hmm, sure looks odd. I counted about 19 of these. `make apidoc` passes. still checking..

File: pkg/virt-controller/leaderelectionconfig/config.go (hunk 2 above)
Comment (RamLavi) [reply pkg/virt-controller/leaderelectionconfig/config.go:42]: Well evidently, this change made it so that the leader info is kept on the lease. 
Fixed accordingly.
Added a comment on the commit desc.

File: api/openapi-spec/swagger.json (hunk 4 above)
Comment (RamLavi) [reply api/openapi-spec/swagger.json:135]: I tend to think it's OK. See [here](https://swagger.io/docs/specification/using-ref/) . 
When I look at the output with the IDE, I don't see an issue though admittedly I am no expert.
@oshoval - WDYT? Does the output look OK to you now?

Comment (oshoval) [reply api/openapi-spec/swagger.json:135]: If IDE parse it then it sounds good
I am not familiar with it, better to ask someone who does
thanks

**Review by jean-edouard — COMMENTED** [review]
/approve

**Comment by Barakmor1** [comment]
Do you think it would be possible to split it into two commits: one for the manual code change and one for the generated code? It's a bit hard to follow, and I'm having trouble loading all the changes on my browser :)

File: tests/libnet/job/job.go [hunk 5]
```diff
@@ -139,10 +138,14 @@ func WaitForJob(job *batchv1.Job, toSucceed bool, timeout time.Duration) error {
 				break
 			case batchv1.JobFailureTarget:
 				break
+			case batchv1.JobSuccessCriteriaMet:
+				break
```
Comment (Barakmor1) [inline tests/libnet/job/job.go:142]: Can we use `default:` instead of all these cases?

File: tests/libnet/job/job.go [hunk 6]
```diff
@@ -139,10 +138,14 @@ func WaitForJob(job *batchv1.Job, toSucceed bool, timeout time.Duration) error {
...
 			}
 		}
 		return !finish, nil
-	})
+	}
+
+	err := wait.PollUntilContextTimeout(context.Background(), time.Second, timeout, true, poller)
```
Comment (Barakmor1) [inline tests/libnet/job/job.go:148]: Please use:
`Expect(err).ToNot(HaveOccurred(),fmt.Sprintf("job %s timeout reached, status: %+v, err: %v", job.Name, job.Status, err))`

File: tests/libnet/job/job.go (hunk 5 above)
Comment (RamLavi) [reply tests/libnet/job/job.go:142]: TBH I'm not sure what should be the default behavior. That's why I preferred to keep it case specific, as was done before

**Comment by RamLavi** [comment]
[change](https://github.com/kubevirt/kubevirt/compare/f87f7cc8cbd9185016952c942f892fdf08718c39..89b7f2f944fd3eddf1af9147c12d372018ad9c00): Rebase

File: tests/libnet/job/job.go (hunk 6 above)
Comment (RamLavi) [reply tests/libnet/job/job.go:148]: According to [writing-good-e2e-tests manual](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/writing-good-e2e-tests.md#interrupting-tests):
```
Using gomega assertions in helper packages is problematic for two reasons:

The stacktrace associated with the failure starts with the helper unless extra care is take to pass in a stack offset.
Additional explanations for a potential failure must be prepared beforehand and passed in.
```
So I think it's better to return the error. Do you agree?

File: tests/libnet/job/job.go (hunk 5 above)
Comment (Barakmor1) [reply tests/libnet/job/job.go:142]: Umm maybe but i think we can aggregate it to a single case something like
`    case "apple", "banana", "orange":`

File: tests/libnet/job/job.go (hunk 6 above)
Comment (Barakmor1) [reply tests/libnet/job/job.go:148]: Yes, you are right. Sorry, I thought it was a test. :)

File: tests/libnet/job/job.go (hunk 5 above)
Comment (RamLavi) [reply tests/libnet/job/job.go:140]: [DONE](https://github.com/kubevirt/kubevirt/compare/5fd9d33e244eae0c69afbb68bee413ecf94965d0..4c1820d066363634fcf46f8c774d91c42f391509)

**Comment by oshoval** [comment]
atm we have a problem that some IDEs don't index symbols correct
because the IDE try to fetch toolchain 1.22 that doesn't exists
https://github.com/golang/go/issues/61455#issuecomment-2072931029
with this PR it fixes it to be 1.22.0 instead that fixed this weird by product

**Comment by oshoval** [comment]
@RamLavi 
is it fine to try and merge https://github.com/kubevirt/kubevirt/pull/11874 meanwhile please?
it will fix IDEs problems
unless this PR will be merged soon, wdyt ?
Thanks

**Comment by RamLavi** [comment]
> @RamLavi 
> is it fine to try and merge https://github.com/kubevirt/kubevirt/pull/11874 meanwhile please?
> it will fix IDEs problems
> unless this PR will be merged soon, wdyt ?
> Thanks
> 

I don't see a reason why this shouldn't be merged promptly, it's already approved, just needs an lgtm

**Comment by fossedihelm** [comment]
@RamLavi Please provide a release note, I think this PR deserves more than `NONE`

**Comment by RamLavi** [comment]
> @RamLavi Please provide a release note, I think this PR deserves more than `NONE`

ACK, DONE

**Comment by RamLavi** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/4c1820d066363634fcf46f8c774d91c42f391509..2c0cd651225fd7d43df7580b85f4fc7e55660b3a): Rebase

**Comment by fossedihelm** [comment]
@RamLavi Can you check the failures? Looks like they are the same in the recent tests, so maybe they are related to this PR. Thank you

**Comment by RamLavi** [comment]
Change: [Rebase](https://github.com/kubevirt/kubevirt/compare/2c0cd651225fd7d43df7580b85f4fc7e55660b3a..0ed9e2ca85d701e0f838281fbf7aadd4a10915b5)

**Comment by RamLavi** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/0ed9e2ca85d701e0f838281fbf7aadd4a10915b5..f438a034f1f4c43f4245cccce1005df809f50b0b): Squash

**Comment by RamLavi** [comment]
> @RamLavi Can you check the failures? Looks like they are the same in the recent tests, so maybe they are related to this PR. Thank you

Squashed now that the commits were lgtmed seperately, squashed as they can't stand alone. 
@fossedihelm let's see again how it goes.

**Comment by RamLavi** [comment]
Change: [Rebase](https://github.com/kubevirt/kubevirt/compare/6092e9cdfaa5449e646d6af1bd8c10932f061ed0..c39ece488d2e37603ccecd0e43b0edcb4abc2b2a)

**Comment by RamLavi** [comment]
As commits are squashed, removing hold
/hold cancel

File: tests/libnet/job/job.go (hunk 5 above)
Comment (fossedihelm) [reply tests/libnet/job/job.go:140]: golang do not fallthrough through cases if not explicit set. So as is now:
			`case batchv1.JobSuspended`: `<do not break>`
			`case batchv1.JobFailureTarget`: `<do not break>`
			`case batchv1.JobSuccessCriteriaMet`: `<it breaks>`
Everything works the same here since a single break works like an empty statement, but keep in mind that right now, it is working not because of an hypotetical fallthrough.
Also consider that, from https://go.dev/ref/spec#Break_statements, the "break" statement terminates execution of the innermost ["for"](https://go.dev/ref/spec#For_statements), ["switch"](https://go.dev/ref/spec#Switch_statements), or ["select"](https://go.dev/ref/spec#Select_statements) statement within the same function.
IOW we can completely drop the last 3 cases.

**Comment by RamLavi** [comment]
/test pull-kubevirt-unit-test
passes locally, could be flake

**Comment by akrejcir** [comment]
This will make it more convenient to import `client-go` in other projects, thanks.

A question out of curiosity: Is it ok to make kubevirt depend on k8s 1.30, when CI still runs on 1.28 and 1.29? I assumed that we support these older versions. Is the k8s.io code backward compatible?

**Comment by RamLavi** [comment]
> @RamLavi - there looks to be an issue with the sig-monitoring lane on this PR. https://prow.ci.kubevirt.io/pr-history/?org=kubevirt&repo=kubevirt&pr=11840

Lemme take a look
/hold

**Comment by fossedihelm** [comment]
This can fix it: https://github.com/kubevirt/kubevirt/pull/11950

**Comment by RamLavi** [comment]
> This will make it more convenient to import `client-go` in other projects, thanks.
> 
> A question out of curiosity: Is it ok to make kubevirt depend on k8s 1.30, when CI still runs on 1.28 and 1.29? I assumed that we support these older versions. Is the k8s.io code backward compatible?

I don't think it's a problem, as long as we make sure to use GAed features and resources, then it should be backwards compatible.

**Comment by oshoval** [comment]
some cosmetics nits here that can be done for leader election change (your change is right, just some var name changes if you want to take a look please)
https://github.com/kubevirt/kubevirt/pull/10640/commits/f6eaf6259693154344b63985acf62e033488d38b

**Comment by RamLavi** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/ca842bf498816bf3c2ec23562b98aa260ac2c586..342c201e85cf44500702004c723ae3ecce447424): Rebase

**Comment by oshoval** [comment]
if the failures are not related please consider unhold

**Comment by RamLavi** [comment]
> if the failures are not related please consider unhold

I don't think they're related, I see very similar tests fail on other PRs (like https://github.com/kubevirt/kubevirt/pull/11516, see failing [here](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11516/pull-kubevirt-e2e-k8s-1.29-sig-monitoring/1769689013830029312))

/hold cancel

**Comment by brianmcarey** [comment]
Looking at the PR history here - it looks like this will be very unlikely to merge until https://github.com/kubevirt/kubevirt/pull/11950 is merged. 

https://prow.ci.kubevirt.io/pr-history/?org=kubevirt&repo=kubevirt&pr=11840 

The hold should probably stay in place until #11950 is merged or we decide to override the monitoring lane here?

**Comment by oshoval** [comment]
but if you override, wont this flake affect everything once this PR is merged?
otherwise sgtm

**Comment by RamLavi** [comment]
/hold
I may need to bump to v0.30.1

**Comment by RamLavi** [comment]
> /hold I may need to bump to v0.30.1

no need as of now. 
/hold cancel

**Comment by oshoval** [comment]
> Looking at the PR history here - it looks like this will be very unlikely to merge until #11950 is merged.
> 
> https://prow.ci.kubevirt.io/pr-history/?org=kubevirt&repo=kubevirt&pr=11840
> 
> The hold should probably stay in place until #11950 is merged or we decide to override the monitoring lane here?

@RamLavi please consider this

**Comment by RamLavi** [comment]
> > Looking at the PR history here - it looks like this will be very unlikely to merge until #11950 is merged.
> > 
> > https://prow.ci.kubevirt.io/pr-history/?org=kubevirt&repo=kubevirt&pr=11840
> > 
> > The hold should probably stay in place until #11950 is merged or we decide to override the monitoring lane here?
> 
> @RamLavi please consider this
> 

It does look bad.. Lets give #11950 more time to get merged

**Comment by RamLavi** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/342c201e85cf44500702004c723ae3ecce447424..732fe87f18c918db757e1df62989af8366489dd4): Rebase

**Comment by RamLavi** [comment]
update: I tried to mount https://github.com/kubevirt/kubevirt/pull/11950 's commit after I made sure they pass on the original code - and indeed it seems like the test fail with my changes. 
I'm trying to investigate why. 
I don't see an obvious reason though. We are not touching relevant virt-API/virt-operator code that could influence this test, so I'm perplexed. 

@machadovilaca - Can you provide a command I could use in order to scrape the relevant prom rule in order to see the rate that is supposed to fire? I suspect that maybe the expression is outdated..

**Comment by Barakmor1** [comment]
> update: I tried to mount #11950 's commit after I made sure they pass on the original code - and indeed it seems like the test fail with my changes. I'm trying to investigate why. I don't see an obvious reason though. We are not touching relevant virt-API/virt-operator code that could influence this test, so I'm perplexed.
> 
> @machadovilaca - Can you provide a command I could use in order to scrape the relevant prom rule in order to see the rate that is supposed to fire? I suspect that maybe the expression is outdated..

@RamLavi  Which test are you referring to?

**Comment by RamLavi** [comment]
> > update: I tried to mount #11950 's commit after I made sure they pass on the original code - and indeed it seems like the test fail with my changes. I'm trying to investigate why. I don't see an obvious reason though. We are not touching relevant virt-API/virt-operator code that could influence this test, so I'm perplexed.
> > @machadovilaca - Can you provide a command I could use in order to scrape the relevant prom rule in order to see the rate that is supposed to fire? I suspect that maybe the expression is outdated..
> 
> @RamLavi Which test are you referring to?

`[Serial][sig-monitoring]Component Monitoring Errors metrics VirtOperatorRESTErrorsBurst and VirtOperatorRESTErrorsHigh should be triggered when requests to virt-operator are failing` ([here](https://github.com/kubevirt/kubevirt/blob/24c6eaa222c75002550270f2f72a174276df4692/tests/monitoring/component_monitoring.go#L213))

**Comment by RamLavi** [comment]
> > > update: I tried to mount #11950 's commit after I made sure they pass on the original code - and indeed it seems like the test fail with my changes. I'm trying to investigate why. I don't see an obvious reason though. We are not touching relevant virt-API/virt-operator code that could influence this test, so I'm perplexed.
> > > @machadovilaca - Can you provide a command I could use in order to scrape the relevant prom rule in order to see the rate that is supposed to fire? I suspect that maybe the expression is outdated..
> > 
> > 
> > @RamLavi Which test are you referring to?
> 
> `[Serial][sig-monitoring]Component Monitoring Errors metrics VirtOperatorRESTErrorsBurst and VirtOperatorRESTErrorsHigh should be triggered when requests to virt-operator are failing` ([here](https://github.com/kubevirt/kubevirt/blob/24c6eaa222c75002550270f2f72a174276df4692/tests/monitoring/component_monitoring.go#L213))

OK, I think I understand what's happening here.
As part of this PR, we moved to select a leader using a lease.
Now, regarding the test: for the test purposes [we replace the timing of the alerts](https://github.com/kubevirt/kubevirt/blob/f4636fa5cac15bd7d813eb67461a27a85572a031/tests/monitoring/component_monitoring.go#L185) to 1 minute
so it will trigger if 80% failed in the last minute, then [we remove the virt-operator cluster-role-binding](https://github.com/kubevirt/kubevirt/blob/f4636fa5cac15bd7d813eb67461a27a85572a031/tests/monitoring/component_monitoring.go#L223), so virt-operator cannot perform any operations in the cluster
which *should* cause all the requests it receives to fail.

But closer examination of the metrics show that the lease GET/SET still receive code 200 (meaning no error).. 
How the leases requests still work even when the cluster-role- binding is dead - I do not know.. 

image

**Comment by machadovilaca** [comment]
> > > > update: I tried to mount #11950 's commit after I made sure they pass on the original code - and indeed it seems like the test fail with my changes. I'm trying to investigate why. I don't see an obvious reason though. We are not touching relevant virt-API/virt-operator code that could influence this test, so I'm perplexed.
> > > > @machadovilaca - Can you provide a command I could use in order to scrape the relevant prom rule in order to see the rate that is supposed to fire? I suspect that maybe the expression is outdated..
> > > 
> > > 
> > > @RamLavi Which test are you referring to?
> > 
> > 
> > `[Serial][sig-monitoring]Component Monitoring Errors metrics VirtOperatorRESTErrorsBurst and VirtOperatorRESTErrorsHigh should be triggered when requests to virt-operator are failing` ([here](https://github.com/kubevirt/kubevirt/blob/24c6eaa222c75002550270f2f72a174276df4692/tests/monitoring/component_monitoring.go#L213))
> 
> OK, I think I understand what's happening here. As part of this PR, we moved to select a leader using a lease. Now, regarding the test: for the test purposes [we remove the virt-operator cluster-role-binding](https://github.com/kubevirt/kubevirt/blob/f4636fa5cac15bd7d813eb67461a27a85572a031/tests/monitoring/component_monitoring.go#L223), so virt-operator cannot perform any operations in the cluster which _should_ cause all the requests it receives to fail.
> 
> But closer examination of the metrics show that the lease GET/SET still receive code 200 (meaning no error).. I think it's because we have `lease` resource both on the `cluster-role` and the `role`... removing it gave the desired alert to fire. @machadovilaca I updated the test that failed to also remove the role-binding and now it succeeds. Do you agree this is the appropriate course of action? If so, I will set it as soon as your PR is merged
> 
> image

@RamLavi great work

If we just need to also remove the role-binding then it's fine by me. Should we do it in a different PR?

**Comment by RamLavi** [comment]
> > > > > update: I tried to mount #11950 's commit after I made sure they pass on the original code - and indeed it seems like the test fail with my changes. I'm trying to investigate why. I don't see an obvious reason though. We are not touching relevant virt-API/virt-operator code that could influence this test, so I'm perplexed.
> > > > > @machadovilaca - Can you provide a command I could use in order to scrape the relevant prom rule in order to see the rate that is supposed to fire? I suspect that maybe the expression is outdated..
> > > > 
> > > > 
> > > > @RamLavi Which test are you referring to?
> > > 
> > > 
> > > `[Serial][sig-monitoring]Component Monitoring Errors metrics VirtOperatorRESTErrorsBurst and VirtOperatorRESTErrorsHigh should be triggered when requests to virt-operator are failing` ([here](https://github.com/kubevirt/kubevirt/blob/24c6eaa222c75002550270f2f72a174276df4692/tests/monitoring/component_monitoring.go#L213))
> > 
> > 
> > OK, I think I understand what's happening here. As part of this PR, we moved to select a leader using a lease. Now, regarding the test: for the test purposes [we remove the virt-operator cluster-role-binding](https://github.com/kubevirt/kubevirt/blob/f4636fa5cac15bd7d813eb67461a27a85572a031/tests/monitoring/component_monitoring.go#L223), so virt-operator cannot perform any operations in the cluster which _should_ cause all the requests it receives to fail.
> > But closer examination of the metrics show that the lease GET/SET still receive code 200 (meaning no error).. I think it's because we have `lease` resource both on the `cluster-role` and the `role`... removing it gave the desired alert to fire. @machadovilaca I updated the test that failed to also remove the role-binding and now it succeeds. Do you agree this is the appropriate course of action? If so, I will set it as soon as your PR is merged
> > image
> 
> @RamLavi great work
> 
> If we just need to also remove the role-binding then it's fine by me. Should we do it in a different PR?

no. If this PR spoiled the test it should be the one to fix it.

**Comment by RamLavi** [comment]
we [broke](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11840/pull-kubevirt-e2e-k8s-1.29-sig-monitoring/1798328758554333184) the streak!

removing hold
/hold cancel

trying again
/test pull-kubevirt-e2e-k8s-1.29-sig-monitoring

**Comment by RamLavi** [comment]
> we [broke](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11840/pull-kubevirt-e2e-k8s-1.29-sig-monitoring/1798328758554333184) the streak!
> 
> removing hold /hold cancel
> 
> trying again /test pull-kubevirt-e2e-k8s-1.29-sig-monitoring

pull-kubevirt-e2e-k8s-1.29-sig-monitoring failed, but on a different test.
/test pull-kubevirt-e2e-k8s-1.29-sig-monitoring

File: tests/monitoring/component_monitoring.go [hunk 7]
```diff
@@ -191,6 +195,13 @@ var _ = Describe("[Serial][sig-monitoring]Component Monitoring", Serial, decorat
 			if !errors.IsAlreadyExists(err) {
 				util.PanicOnError(err)
 			}
+			operatorRoleBinding.Annotations = nil
+			operatorRoleBinding.ObjectMeta.ResourceVersion = ""
+			operatorRoleBinding.ObjectMeta.UID = ""
+			_, err = virtClient.RbacV1().RoleBindings(flags.KubeVirtInstallNamespace).Create(context.Background(), operatorRoleBinding, metav1.CreateOptions{})
+			if !errors.IsAlreadyExists(err) {
+				Expect(err).ToNot(HaveOccurred())
+			}
```
Comment (fossedihelm) [inline tests/monitoring/component_monitoring.go:204]:
```suggestion
			Expect(err).To(Or(MatchError(errors.IsAlreadyExists, "IsAlreadyExists"), Not(HaveOccurred())))
```
I would align the code also in lines 195-197

Comment (fossedihelm) [reply tests/monitoring/component_monitoring.go]: My bad:
```
Expect(err).To(Or(Not(HaveOccurred(), MatchError(errors.IsAlreadyExists, "IsAlreadyExists"))))
```

Comment (fossedihelm) [reply tests/monitoring/component_monitoring.go]: @RamLavi I think you have to change the Expect as I suggested before retest

Comment (RamLavi) [reply tests/monitoring/component_monitoring.go]: oh I missed that. you're right. will change.

Comment (RamLavi) [reply tests/monitoring/component_monitoring.go]: Changed to:
```
Expect(err).To(Or(Not(HaveOccurred()), MatchError(errors.IsAlreadyExists, "IsAlreadyExists")))
```

**Comment by RamLavi** [comment]
/test pull-kubevirt-unit-test-arm64
looks like a flake

**Comment by oshoval** [comment]
Did you take a look on those https://github.com/kubevirt/kubevirt/pull/11840#issuecomment-2124494884 ?
Just some cosmetics, doesnt merit hold, but nice to have
thanks

**Comment by RamLavi** [comment]
> Did you take a look on those [#11840 (comment)](https://github.com/kubevirt/kubevirt/pull/11840#issuecomment-2124494884) ? Just some cosmetics, doesn't merit hold, but nice to have thanks

Hey @oshoval , thanks for the heads up. I prefer not to hold the PR for this specific change at this point.

File: pkg/virt-api/api.go [hunk 8]
```diff
@@ -786,7 +786,7 @@ func (app *virtAPIApp) Compose() {
 func (app *virtAPIApp) ConfigureOpenAPIService() {
 	config := openapi.CreateConfig()
 	config.GetDefinitions = v12.GetOpenAPIDefinitions
-	spec, err := builderv3.BuildOpenAPISpecFromRoutes(restfuladapter.AdaptWebServices(restful.RegisteredWebServices()), config)
+	spec, err := builder.BuildOpenAPISpecFromRoutes(restfuladapter.AdaptWebServices(restful.RegisteredWebServices()), config)
```
Comment (xpivarc) [inline pkg/virt-api/api.go:789]: Why do we roll back?

File: pkg/virt-controller/leaderelectionconfig/config.go (hunk 2 above)
Comment (xpivarc) [reply pkg/virt-controller/leaderelectionconfig/config.go:42]: Thanks, makes sense.

**Review by xpivarc — COMMENTED** [review]
There is one more thing that I am not sure about. The first commit could be split but no requirement

**Comment by RamLavi** [comment]
> There is one more thing that I am not sure about. The first commit could be split but no requirement

I split it at the time to be more readable but eventually we must squash because the changes don't compile otherwise..

File: pkg/virt-api/api.go (hunk 8 above)
Comment (RamLavi) [reply pkg/virt-api/api.go:789]: Once I bumped the k8s version there was a version inconsistency, and `config` (of type `*common.Config`, created by `openapi.CreateConfig()`) was no longer of the appropriate type (`*common.OpenAPIV3Config`).

They moved the config to the v3 API.

Anyways per your question - the reason is I thought it would be the least change to move to the old function to keep consistency, as changing to builderv3 will result in other bigger changes.

But if you think that we should bump swagger to v3 as well then we can do that.

**Comment by RamLavi** [comment]
I think this OR is made redundant with https://github.com/kubevirt/kubevirt/pull/12097 merged.
Closing
